test/functional/feature_asmap.py

```diff
@@ -133,7 +133,8 @@ def run_test(self):
         self.node = self.nodes[0]
         self.datadir = self.node.chain_path
         self.default_asmap = os.path.join(self.datadir, DEFAULT_ASMAP_FILENAME)
-        self.asmap_raw = os.path.join(os.path.dirname(os.path.realpath(__file__)), ASMAP)
+        base_dir = self.config["environment"]["SRCDIR"]
+        self.asmap_raw = os.path.join(base_dir, ASMAP)
```

Comment:
Other cases point to files that reside in the build directory. Or do you have any particular example?

Reply:
Can our previous [conversation](https://github.com/hebasto/bitcoin/pull/226#discussion_r1632362329) be helpful?

Reply:
Please disregard my previous comments if they contradict this one. My apologies for any confusion.

On Windows, creating symbolic links requires a [permission](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-10/security/threat-protection/security-policy-settings/create-symbolic-links), which is disabled by default. Therefore, CMake creates hard links (or copies as a fallback). In that case, `os.path.realpath(__file__)` will point at a file in the build directory.

An example of an error can be observed [here](https://github.com/hebasto/bitcoin/actions/runs/9972592403/job/27556158771):
```
268/313 - feature_asmap.py failed, Duration: 2 s
stdout:
2024-07-17T11:26:07.749000Z TestFramework (INFO): PRNG seed is: 7216823379970217686
2024-07-17T11:26:07.780000Z TestFramework (INFO): Initializing test directory D:\a\_temp\test_runner_₿_🏃_20240717_110914\feature_asmap_42
2024-07-17T11:26:08.584000Z TestFramework (INFO): Test bitcoind with no -asmap arg passed
2024-07-17T11:26:09.525000Z TestFramework (INFO): Test bitcoind -asmap=<absolute path>
2024-07-17T11:26:09.629000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "D:\a\bitcoin\bitcoin\build\test\functional\test_framework\test_framework.py", line 132, in main
    self.run_test()
  File "D:\a\bitcoin\bitcoin\build\test\functional\feature_asmap.py", line 139, in run_test
    self.test_asmap_with_absolute_path()
  File "D:\a\bitcoin\bitcoin\build\test\functional\feature_asmap.py", line 61, in test_asmap_with_absolute_path
    shutil.copyfile(self.asmap_raw, filename)
  File "C:\hostedtoolcache\windows\Python\3.12.4\x64\Lib\shutil.py", line 260, in copyfile
    with open(src, 'rb') as fsrc:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'D:\\a\\bitcoin\\bitcoin\\build\\test\\functional\\../../src/test/data/asmap.raw'
```

Reply:
> Ok, so it can be reproduced in Linux with hard links.

Right. But we do not use hard links on non-Windows systems.

> However, this leads back to my original question: Why not change the other places as well?

In other places, `os.path.realpath(__file__)` is used in conjunction with relative paths in the `test/functional/` directory. Soft links (on non-Windows systems) and hard links (on Windows) are created for every file in the `test/functional/` directory recursively. Therefore, the created paths are valid.

`feature_asmap.py` is an exception because `data/asmap.raw` resides outside of the `test/functional/` directory.

As an alternative, the `data/asmap.raw` file can be moved to somewhere within the `test/functional/` directory.

Reply:
Your example creates a hard link on Linux, which is not what happens in the CMake-based build system.

Do you think we should adjust our functional tests in way to handle hard links on Linux?